src/testing/io.zig

```diff
@@ -0,0 +1,229 @@
+                *anyopaque => @intFromPtr(fd) - 1,
+                else => unreachable,
+            }
+        ];
+    }
+
+    /// Return true with probability x/100.
+    fn x_in_100(io: *IO, x: u8) bool {
+        assert(x <= 100);
+        return x > io.prng.random().uintLessThan(u8, 100);
+    }
+
+    fn submit(
+        self: *IO,
+        context: anytype,
+        comptime callback: anytype,
+        completion: *Completion,
+        comptime operation_tag: std.meta.Tag(Operation),
+        operation_data: anytype,
+        comptime OperationImpl: type,
+    ) void {
+        const onCompleteFn = struct {
+            fn onComplete(io: *IO, _completion: *Completion) void {
+                // Perform the actual operation
+                const op_data = &@field(_completion.operation, @tagName(operation_tag));
+                const result = OperationImpl.do_operation(io, op_data);
+
+                // Complete the Completion
+
+                return callback(
+                    @ptrCast(@alignCast(_completion.context)),
+                    _completion,
+                    result,
+                );
+            }
+        }.onComplete;
+
+        completion.* = .{
+            .next = null,
+            .context = context,
+            .callback = onCompleteFn,
+            .operation = @unionInit(Operation, @tagName(operation_tag), operation_data),
+        };
+
+        self.completed.push(completion);
+    }
+
+    pub const ReadError = error{
+        WouldBlock,
+        NotOpenForReading,
+        ConnectionResetByPeer,
+        Alignment,
+        InputOutput,
+        IsDir,
+        SystemResources,
+        Unseekable,
+        ConnectionTimedOut,
+    } || posix.UnexpectedError;
+
+    pub fn read(
+        self: *IO,
+        comptime Context: type,
+        context: Context,
+        comptime callback: fn (
+            context: Context,
+            completion: *Completion,
+            result: ReadError!usize,
+        ) void,
+        completion: *Completion,
+        fd: posix.fd_t,
+        buffer: []u8,
+        offset: u64,
+    ) void {
+        self.submit(
+            context,
+            callback,
+            completion,
+            .read,
+            .{
+                .fd = fd,
+                .buf = buffer.ptr,
+                .len = @as(u32, @intCast(buffer_limit(buffer.len))),
+                .offset = offset,
+            },
+            struct {
+                fn do_operation(io: *IO, op: anytype) ReadError!usize {
+                    if (io.x_in_100(io.options.read_fault_probability) or
+                        (op.len > constants.sector_size and
+                        io.x_in_100(io.options.larger_than_logical_sector_read_fault_probability)))
+                    {
+                        return error.InputOutput;
+                    }
+
+                    const data = io.get_file_data(op.fd);
+                    @memcpy(op.buf[0..op.len], data[op.offset..][0..op.len]);
```

Comment:
Prefer [stdx.copy_disjoint](https://github.com/tigerbeetle/tigerbeetle/pull/2031) (I may have been the reason for DJ writing that PR :joy:).